### ISSUE TITLE
cache-and-network clears previous data when refetching

### ISSUE THREAD

**Issue by zenios** [FormidableLabs/urql-exchange-graphcache#162]
Using cache-and-network after initial fetching i get data with stale true followed by undefined data and then again data with stale false

Wasn't cache-and-network supposed to never return undefined for data unless its and initial fetch?

**Comment by JoviDeCroock** [comment]
Hey @zenios 

I think there's some context missing in this issue report, what is your response? Can you give a reproduction? Are there any resolvers that could interfere?

**Comment by zenios** [comment]
I did some tests and found out that this happens only when we have dedupExchange in place.

Should dedupExchange be ignored when using graphcache?

**Comment by JoviDeCroock** [comment]
The `dedupExchange` should not  be an issue, I think there's a more underlying issue being masked. It would really be helpfull to see as to why you're using `undefined` and what's going wrong.
Removing `dedup` could point to an operation eagerly returning during `cache-and-network` or multiple dispatches at the same time removing this?

**Comment by zenios** [comment]
I am not using undefined. Data is undefined such as 

[{fetching, data, stale}] = useQuery(params)

First mount
data = undefined  | fetching true | stale = false
data = Some data | fetching false | stale = false

Second mount

data = some data | fetching false | stale = true
data = undefined | fetching true | stale = false
data = new data | fetching false | state = false


I found something else also.

This happens only when executing in StrictMode in combination with dedupExchange

My findings are

StrictMode WITHOUT dedupExchange

Everything works (No undefined data on second mount) but queries execute twice (Two requests per mount on the server)

StrictMode AND dedupExchange

Undefined as data on second mount (Only one request per mount on the server)

Non StrictMode AND dedupExchnage

Everything works (No undefined data on second mount) and queries execute only once

Non StrictMode WITHOUT dedupExchange

Everything works (No undefined data on second mount) and queries execute only once

**Comment by zenios** [comment]
Forgot to mention

Urql: 1.8.2
exchange-graphcache: 2.1.0

Do you

**Comment by JoviDeCroock** [comment]
Would you mind trying with `yarn add https://github.pika.dev/formidablelabs/urql/pr/521`?

**Comment by zenios** [comment]
Same results

**Comment by zenios** [comment]
The only difference with the previous results is that it does the undefined data also on non StrictMode

**Comment by JoviDeCroock** [comment]
So yes the thing we can conclude is that it's not `graphCache` related but the combination of `urql-operations` with `StrictMode` that's good to know. I'm going to look into this.

**Comment by zenios** [comment]
Do you have a codesandbox example with graphcache?

Maybe i can work on that and create a reproduction

**Comment by zenios** [comment]
Also i know that there are some changes with regards to strictmode after 1.8.2 but they are not released. If you could create a @next version i can try that also

**Comment by JoviDeCroock** [comment]
You can find a repro template here: https://raw.githubusercontent.com/FormidableLabs/urql/master/.github/ISSUE_TEMPLATE/bug_report.md

**Comment by zenios** [comment]
I have a reproduction. Even without graphcache

https://codesandbox.io/s/recursing-bose-ijmri?fontsize=14&hidenavigation=1&theme=dark

The examples shows loading text only when data is undefined. You can see clearly when toggling between component show / hide that is loading.

**Comment by JoviDeCroock** [comment]
When using the build from `yarn add https://github.pika.dev/formidablelabs/urql/pr/521` the bug you describe doesn't pop up for me

**Comment by zenios** [comment]
Let me try again after clearing node modules

**Comment by zenios** [comment]
Still the same for me.Have you tried the PR on the codesandbox example i provided?

**Comment by JoviDeCroock** [comment]
Yes, where it is fixed with that PR

**Comment by zenios** [comment]
That reproduction though does not include graphcache

**Comment by zenios** [comment]
Most probably we are looking at different results.I forked the codesandbox project and added the version of urql specified in the PR and i still see the issue.

**Comment by JoviDeCroock** [comment]
@zenios I added it to my local reproduction and it works even with graphcache.

Here's a gif with graphcache and that urql pr: http://www.giphy.com/gifs/M9TYuB6jf3qFZz295x you'll see fetching in the column if ever `res.fetching || res.data === undefined`

**Comment by zenios** [comment]
Actually the reproduction will show loading only if data === undefined.
I removed the res.fetching just to show that on remount we should always
have data (cache-and-network) even though its fetching

Btw: Your Giphy returns 404

**Comment by zenios** [comment]
I am positive that this is not the case. Do you want me to make a video as well?

**Comment by JoviDeCroock** [comment]
I found the caveat and I'm researching it

**Comment by zenios** [comment]
Ok thanks

**Comment by JoviDeCroock** [comment]
Hey,

I think this `yarn add https://github.pika.dev/formidablelabs/urql/pr/523` solves your issue

**Comment by zenios** [comment]
Will try in the next 30 minutes and let you know

Thanks

**Comment by zenios** [comment]
Hi 

So i have good new and bad news :)

The good new is that we don't have any more undefined data between refetches in strict mod

The bad news is that fetching is not raised to true when re fetching.

It stays false (Even though a re-fetch is actually executed)

**Comment by JoviDeCroock** [comment]
For a background fetch? Yes that's normal the stale indicate that a background fetch is going on

**Comment by zenios** [comment]
I though fetching was supposed to also raise to true. (Just like apollo)

Never the less stale should be fine 

Thank you very much for your help. 
Do you want me to close the issue or wait until the code is merged upstream?

**Comment by JoviDeCroock** [comment]
I'll be closing this, feel free to open a new issue if you walk into other issues.
Have a nice day!
